File: scripts/keep-alive.py
#!/usr/bin/env python3
"""
Keep-Alive Ping Script for Render Free Tier (Python Version)

Gửi request ping định kỳ đến FastAPI server để tránh bị sleep
Hỗ trợ retry, logging, và xử lý lỗi

Usage: python keep-alive.py <server_url> [retries] [timeout]
Ví dụ: python keep-alive.py https://your-app.onrender.com 3 10
"""

import sys
import os
import time
import requests
from datetime import datetime
from typing import Dict, Optional
from urllib.parse import urljoin

# ============ CONFIG ============
CONFIG = {
    'SERVER_URL': os.environ.get('RENDER_URL') or (sys.argv[1] if len(sys.argv) > 1 else 'https://your-render-url.onrender.com'),
    'HEALTH_ENDPOINT': '/health',
    'MAX_RETRIES': int(sys.argv[2]) if len(sys.argv) > 2 else int(os.environ.get('PING_RETRIES', 3)),
    'TIMEOUT_SECONDS': int(sys.argv[3]) if len(sys.argv) > 3 else int(os.environ.get('PING_TIMEOUT', 10)),
    'VERBOSE': os.environ.get('VERBOSE') == 'true' or '--verbose' in sys.argv,
}

# ============ LOGGING ============
class Logger:
    @staticmethod
    def _timestamp():
        return datetime.now().isoformat()

    @staticmethod
    def log(msg: str):
        print(f"[{Logger._timestamp()}] ℹ️  {msg}")

    @staticmethod
    def success(msg: str):
        print(f"[{Logger._timestamp()}] ✅ {msg}")

    @staticmethod
    def error(msg: str):
        print(f"[{Logger._timestamp()}] ❌ {msg}", file=sys.stderr)

    @staticmethod
    def warn(msg: str):
        print(f"[{Logger._timestamp()}] ⚠️  {msg}", file=sys.stderr)

    @staticmethod
    def debug(msg: str):
        if CONFIG['VERBOSE']:
            print(f"[{Logger._timestamp()}] 🔍 {msg}")

# ============ PING LOGIC ============
def ping_server(url: str, timeout: int = CONFIG['TIMEOUT_SECONDS']) -> Dict:
    """
    Gửi GET request đến health endpoint
    
    Returns:
        Dict: {success, status_code, message, duration}
    """
    start_time = time.time()
    
    try:
        response = requests.get(url, timeout=timeout)
        duration = round((time.time() - start_time) * 1000, 2)  # ms
        
        if response.status_code == 200:
            return {
                'success': True,
                'status_code': response.status_code,
                'message': 'Server responded with 200 OK',
                'duration': duration,
            }
        else:
            return {
                'success': False,
                'status_code': response.status_code,
                'message': f'Unexpected status code: {response.status_code}',
                'duration': duration,
            }
    except requests.exceptions.Timeout:
        duration = round((time.time() - start_time) * 1000, 2)
        return {
            'success': False,
            'status_code': 0,
            'message': f'Request timeout after {timeout}s',
            'duration': duration,
        }
    except requests.exceptions.RequestException as e:
        duration = round((time.time() - start_time) * 1000, 2)
        return {
            'success': False,
            'status_code': 0,
            'message': f'Request error: {str(e)}',
            'duration': duration,
        }

def ping_with_retry(url: str, max_retries: int = CONFIG['MAX_RETRIES']) -> Dict:
    """
    Retry logic với exponential backoff
    
    Returns:
        Dict: {success, attempts, last_result}
    """
    last_result = None
    consecutive_failures = 0

    for attempt in range(1, max_retries + 1):
        try:
            Logger.debug(f"Attempt {attempt}/{max_retries}")
            last_result = ping_server(url)

            if last_result['success']:
                Logger.success(
                    f"Server is alive ({last_result['status_code']}) - "
                    f"{last_result['duration']}ms - Attempt {attempt}/{max_retries}"
                )
                return {
                    'success': True,
                    'attempts': attempt,
                    'last_result': last_result,
                }

            consecutive_failures += 1
            Logger.warn(
                f"Attempt {attempt} failed: {last_result['message']} "
                f"({last_result['duration']}ms)"
            )

            # Retry với delay exponential backoff (1s, 2s, 4s)
            if attempt < max_retries:
                delay_seconds = 2 ** (attempt - 1)
                Logger.debug(f"Waiting {delay_seconds}s before retry...")
                time.sleep(delay_seconds)

        except Exception as e:
            consecutive_failures += 1
            Logger.error(f"Attempt {attempt} error: {str(e)}")

    # Tất cả attempts thất bại
    Logger.error(
        f"❌ All {max_retries} attempts failed. "
        f"Last error: {last_result['message'] if last_result else 'Unknown error'}"
    )

    # Log cảnh báo nếu fail quá nhiều lần
    if consecutive_failures >= max_retries:
        Logger.warn(
            f"⚠️  ALERT: Server may be down after {consecutive_failures} consecutive failures"
        )

    return {
        'success': False,
        'attempts': max_retries,
        'last_result': last_result,
    }

# ============ MAIN ============
# ============ MAIN ============
def main():
    # Nếu URL đã có /health hoặc endpoint khác, dùng luôn; nếu không, thêm /health
    if CONFIG['SERVER_URL'].endswith('/'):
        url = CONFIG['SERVER_URL'].rstrip('/') + CONFIG['HEALTH_ENDPOINT']
    elif '/health' in CONFIG['SERVER_URL'] or '/status' in CONFIG['SERVER_URL']:
        url = CONFIG['SERVER_URL']
    else:
        url = CONFIG['SERVER_URL'] + CONFIG['HEALTH_ENDPOINT']
    
    Logger.log("Starting keep-alive ping...")
    Logger.log(f"Target: {url}")
    Logger.log(f"Max retries: {CONFIG['MAX_RETRIES']}")
    Logger.log(f"Timeout: {CONFIG['TIMEOUT_SECONDS']}s")

    result = ping_with_retry(url)

    # Exit code: 0 = success, 1 = failure
    sys.exit(0 if result['success'] else 1)

if __name__ == '__main__':
    main()

File: scripts/keep-alive.js
#!/usr/bin/env node
/**
 * Keep-Alive Ping Script for Render Free Tier
 * 
 * Gửi request ping định kỳ đến FastAPI server để tránh bị sleep
 * Hỗ trợ retry, logging, và xử lý lỗi
 * 
 * Usage: node keep-alive.js <server_url> [retries] [timeout_ms]
 * Ví dụ: node keep-alive.js https://your-app.onrender.com 3 10000
 */

const https = require('https');
const http = require('http');

// ============ CONFIG ============
const CONFIG = {
  // Lấy từ environment variable hoặc command line
  SERVER_URL: process.env.RENDER_URL || process.argv[2] || 'https://your-render-url.onrender.com',
  HEALTH_ENDPOINT: '/health',
  MAX_RETRIES: parseInt(process.argv[3]) || parseInt(process.env.PING_RETRIES) || 3,
  TIMEOUT_MS: parseInt(process.argv[4]) || parseInt(process.env.PING_TIMEOUT) || 10000,
  VERBOSE: process.env.VERBOSE === 'true' || process.argv.includes('--verbose'),
};

// ============ LOGGING ============
const logger = {
  log: (msg) => console.log(`[${new Date().toISOString()}] ℹ️  ${msg}`),
  success: (msg) => console.log(`[${new Date().toISOString()}] ✅ ${msg}`),
  error: (msg) => console.error(`[${new Date().toISOString()}] ❌ ${msg}`),
  warn: (msg) => console.warn(`[${new Date().toISOString()}] ⚠️  ${msg}`),
  debug: (msg) => CONFIG.VERBOSE && console.log(`[${new Date().toISOString()}] 🔍 ${msg}`),
};

// ============ PING LOGIC ============
/**
 * Gửi request đến health endpoint
 * @returns {Promise<{success: boolean, statusCode: number, message: string, duration: number}>}
 */
function pingServer(url, timeout = CONFIG.TIMEOUT_MS) {
  return new Promise((resolve) => {
    const startTime = Date.now();
    const protocol = url.startsWith('https') ? https : http;

    const request = protocol.get(url, { timeout }, (res) => {
      const duration = Date.now() - startTime;
      
      if (res.statusCode === 200) {
        resolve({
          success: true,
          statusCode: res.statusCode,
          message: `Server responded with 200 OK`,
          duration,
        });
      } else {
        resolve({
          success: false,
          statusCode: res.statusCode,
          message: `Unexpected status code: ${res.statusCode}`,
          duration,
        });
      }
    });

    request.on('timeout', () => {
      request.destroy();
      const duration = Date.now() - startTime;
      resolve({
        success: false,
        statusCode: 0,
        message: `Request timeout after ${timeout}ms`,
        duration,
      });
    });

    request.on('error', (error) => {
      const duration = Date.now() - startTime;
      resolve({
        success: false,
        statusCode: 0,
        message: `Request error: ${error.message}`,
        duration,
      });
    });
  });
}

/**
 * Retry logic với exponential backoff
 * @returns {Promise<{success: boolean, attempts: number, lastResult: object}>}
 */
async function pingWithRetry(url, maxRetries = CONFIG.MAX_RETRIES) {
  let lastResult = null;
  let consecutiveFailures = 0;

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      logger.debug(`Attempt ${attempt}/${maxRetries}`);
      lastResult = await pingServer(url);

      if (lastResult.success) {
        logger.success(
          `Server is alive (${lastResult.statusCode}) - ${lastResult.duration}ms - Attempt ${attempt}/${maxRetries}`
        );
        return {
          success: true,
          attempts: attempt,
          lastResult,
        };
      }

      consecutiveFailures++;
      logger.warn(`Attempt ${attempt} failed: ${lastResult.message} (${lastResult.duration}ms)`);

      // Retry với delay exponential backoff (1s, 2s, 4s)
      if (attempt < maxRetries) {
        const delayMs = Math.pow(2, attempt - 1) * 1000;
        logger.debug(`Waiting ${delayMs}ms before retry...`);
        await new Promise(resolve => setTimeout(resolve, delayMs));
      }
    } catch (error) {
      consecutiveFailures++;
      logger.error(`Attempt ${attempt} error: ${error.message}`);
    }
  }

  // Tất cả attempts thất bại
  logger.error(
    `❌ All ${maxRetries} attempts failed. Last error: ${lastResult?.message || 'Unknown error'}`
  );

  // Log cảnh báo nếu fail quá nhiều lần
  if (consecutiveFailures >= maxRetries) {
    logger.warn(`⚠️  ALERT: Server may be down after ${consecutiveFailures} consecutive failures`);
  }

  return {
    success: false,
    attempts: maxRetries,
    lastResult,
  };
}

// ============ MAIN ============
async function main() {
  // Nếu URL đã có /health hoặc endpoint khác, dùng luôn; nếu không, thêm /health
  const url = CONFIG.SERVER_URL.endsWith('/') 
    ? CONFIG.SERVER_URL.slice(0, -1) + CONFIG.HEALTH_ENDPOINT
    : CONFIG.SERVER_URL.includes('/health') || CONFIG.SERVER_URL.includes('/status')
    ? CONFIG.SERVER_URL
    : CONFIG.SERVER_URL + CONFIG.HEALTH_ENDPOINT;
  
  logger.log(`Starting keep-alive ping...`);
  logger.log(`Target: ${url}`);
  logger.log(`Max retries: ${CONFIG.MAX_RETRIES}`);
  logger.log(`Timeout: ${CONFIG.TIMEOUT_MS}ms`);

  const result = await pingWithRetry(url);

  // Exit code: 0 = success, 1 = failure
  process.exit(result.success ? 0 : 1);
}

// ============ RUN ============
main().catch((error) => {
  logger.error(`Unexpected error: ${error.message}`);
  process.exit(1);
});
